Add functionality for more customized OOC synthesis runs

ganoam opened the pull request:
- Add field to specify one or more TCL scripts to run prior to each individual step
- Add fields to add custom command options to each step.
- Create snapshots of all used source files

ganoam committed:
Add functionality for more customized OOC synthesis runs

- Add field to specify one or more TCL scripts to run prior to each
  individual step
- Add fields to add custom command options to each step.
- Create snapshots of all used source files

ganoam commented:
This PR adds additional fields to the OOC synthesis target:
- `PreStepTcl []core.Path 
- `StepOptions []string`
where `Step` in `{Synth, Opt, Place, PhysOpt, Route}`

PreStepTCL:
Specify a list of TCL scripts to run prior to calling the step command
Example Usage:
```
var synth = xilinx.SynthOutOfContext{
  [...]
  PrePlaceTcl: ins("scripts/place_cores.tcl"),
  RouteDirective: []string{"
    -directive explore", 
    "-foo bar"
```
Will source the script `place_core.tcl` prior to calling `place_design`, and call route_design with the specified flags.

Additionally, I am creating a new subdirectory in `synth_reports` collecting snapshots of all sources used for this run.

exp-sota commented:
The pre-step tcl idea looks useful.

A quick thought. After reading a xilinx doc, it might be better to:
1. be able to switch on and off `phys_opt_design` based on one's needs;
2. be able to choose, for example, either a single `phys_opt_design` run or iterative `phys_opt_design` loop with convergence threshold.
What do you think?

ganoam committed:
- Change to add step TCL fields. If none are supplied, default synthesis is performed
- Add ReportTcl field to support custom reporting

ganoam commented:
As discussed in person, this was much too rigid.
Trying again.

Adding fields to supply custom scripts for each step.
If none is supplied, the default step is executed
Additionally, adding the following fields:

- `PrepareTcl`: Can be optionally supplied to be run after reading source files and constraints and before the synthesis step.
  Intended to set up any design / run variables which could be used by later steps, e.g. to turn on / off reporting for individual steps.
- `ReportTcl` a field to supply a custom reporting script to replace the default reports. Additionally, the reporting code can be individually enabled / disabled by setting the variable `$step_reporting`. E.g. in the `PrepareTcl` script (default: all on).

ganoam committed:
- Go fmt on tgo files
- Re-format template to use tabs instead of spaces (as original does).

ganoam committed:
Add separate fields for timing and different constraints

- Custom timing constraints to override default auto generated timing constraints
- Additional constraints to set e.g. PBblocks and more.
